machine-runner/src/runner/runner.ts

```diff
@@ -84,14 +88,34 @@ export type MachineRunner<SwarmProtocolName extends string, MachineName extends
    * }
    * machine.isDestroyed() // returns false
    */
-  noAutoDestroy: () => MachineRunnerIterableIterator<SwarmProtocolName, MachineName>
-} & MachineRunnerIterableIterator<SwarmProtocolName, MachineName>
+  noAutoDestroy: () => MachineRunnerIterableIterator<SwarmProtocolName, MachineName, StateUnion>
+
+  /**
+   * Add type refinement to the state payload produced by the machine-runner
```

Comment:
Correctly using this function requires passing ALL possible states, right? In that case, we should probably document this fact and check it at runtime (i.e. throw an exception if a state is missing).

Reply:
Currently, it doesn't. That's a good constraint thought, I'll add that.